remotes/fixup.go

```diff
@@ -136,3 +179,55 @@ func checkBaseImage(baseImage *bundle.BaseImage) error {
 
 	return nil
 }
+
+// FixupEvent is an event that is raised by the Fixup Logic
+type FixupEvent struct {
+	SourceImage    string
+	DestinationRef reference.Named
+	Descriptor     *ocischemav1.Descriptor
+	EventType      FixupEventType
+	Message        string
+	Error          error
+	DigestTree     *DigestTree
+}
+
+// FixupEventType is the the type of event raised by the Fixup logic
+type FixupEventType string
+
+const (
+	// FixupEventTypeCopyImageStart is raised when the Fixup logic starts copying an
+	// image
+	FixupEventTypeCopyImageStart = FixupEventType("CopyImageStart")
+
+	// FixupEventTypeCopyImageEnd is raised when the Fixup logic stops copying an
+	// image. Error might be populated
+	FixupEventTypeCopyImageEnd = FixupEventType("CopyImageEnd")
+
+	// FixupEventTypeMountDescriptorStart is raised when the Fixup logic starts
+	// mounting a blob
+	FixupEventTypeMountDescriptorStart = FixupEventType("MountDescriptorStart")
+
+	// FixupEventTypeMountDescriptorEnd is raised when the Fixup logic stops
+	// mounting a blob. Error might be populated
+	FixupEventTypeMountDescriptorEnd = FixupEventType("MountDescriptorEnd")
+
+	// FixupEventTypeCopyDescriptorStart is raised when the Fixup logic starts
+	// copying a blob or manifest
+	FixupEventTypeCopyDescriptorStart = FixupEventType("CopyDescriptorStart")
+
+	// FixupEventTypeCopyDescriptorEnd is raised when the Fixup logic stops
+	// copying a blob or manifest. Error might be populated
+	FixupEventTypeCopyDescriptorEnd = FixupEventType("CopyDescriptorEnd")
+
+	// FixupEventTypeSkipDescriptorPresent is raised when the Fixup logic skips a
+	// descriptor because it is already present in the target repository
+	FixupEventTypeSkipDescriptorPresent = FixupEventType("SkipDescriptorPresent")
+
+	// FixupEventTypeSkipDescriptorForeign is raised when the Fixup logic skips a
+	// descriptor because it references a foreign layer
+	FixupEventTypeSkipDescriptorForeign = FixupEventType("SkipDescriptorForeign")
+
+	// FixupEventTypeDigestTreeResolved is raised when Fixup logic has finished
+	// resolving the descriptor trees to copy/mount
+	FixupEventTypeDigestTreeResolved = FixupEventType("DigestTreeResolved")
```

Comment:
Most other things have a Start/Stop pair, why not this one?

Reply:
Because DigestTree resolution is actually the first step of CopyImage

Reply:
I think this (and my comment about correlation) I mostly doc issues, i.e. which things form sequences, which nest inside each other etc.